File: omni/extensions/runtime/source/omni.usdphysics.tests/python/scripts/tests/attachment.py
# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: BSD-3-Clause
#
import carb
from .base import UsdPhysicsBaseTest


class UsdPhysicsAttachmentTest(UsdPhysicsBaseTest):

    async def test_attachment(self):
        self.fail_on_log_error = True
        self.expected_prims = {}

        vtxvtx_dict = {}
        vtxvtx_dict["enabled"] = False
        vtxvtx_dict["src0"] = "/World/Actor0"
        vtxvtx_dict["src1"] = "/World/Actor1"
        vtxvtx_dict["stiffness"] = float(1.0)
        vtxvtx_dict["damping"] = float(10.0)
        self.expected_prims["/World/VtxVtxAttachment" + "/vtxVtxAttachment"] = vtxvtx_dict

        vtxtri_dict = {}
        vtxtri_dict["enabled"] = True
        vtxtri_dict["src0"] = "/World/Actor0"
        vtxtri_dict["src1"] = "/World/Actor1"
        vtxtri_dict["stiffness"] = float(1.0)
        vtxtri_dict["damping"] = float(10.0)
        self.expected_prims["/World/VtxTriAttachment" + "/vtxTriAttachment"] = vtxtri_dict

        vtxtet_dict = {}
        vtxtet_dict["enabled"] = True
        vtxtet_dict["src0"] = "/World/Actor0"
        vtxtet_dict["src1"] = "/World/Actor1"
        vtxtet_dict["stiffness"] = float(1.0)
        vtxtet_dict["damping"] = float(10.0)
        self.expected_prims["/World/VtxTetAttachment" + "/vtxTetAttachment"] = vtxtet_dict

        vtxcrv_dict = {}
        vtxcrv_dict["enabled"] = True
        vtxcrv_dict["src0"] = "/World/Actor0"
        vtxcrv_dict["src1"] = "/World/Actor1"
        vtxcrv_dict["stiffness"] = float(1.0)
        vtxcrv_dict["damping"] = float(10.0)
        self.expected_prims["/World/VtxCrvAttachment" + "/vtxCrvAttachment"] = vtxcrv_dict

        vtxxform_dict = {}
        vtxxform_dict["enabled"] = True
        vtxxform_dict["src0"] = "/World/Actor0"
        vtxxform_dict["src1"] = "/World/Actor1"
        vtxxform_dict["stiffness"] = float(1.0)
        vtxxform_dict["damping"] = float(10.0)
        self.expected_prims["/World/VtxXformAttachment" + "/vtxXformAttachment"] = vtxxform_dict

        tetxform_dict = {}
        tetxform_dict["enabled"] = True
        tetxform_dict["src0"] = "/World/Actor0"
        tetxform_dict["src1"] = "/World/Actor1"
        tetxform_dict["stiffness"] = float(1.0)
        tetxform_dict["damping"] = float(10.0)
        self.expected_prims["/World/TetXformAttachment" + "/tetXformAttachment"] = tetxform_dict

        tritri_dict = {}
        tritri_dict["enabled"] = True
        tritri_dict["src0"] = "/World/Actor0"
        tritri_dict["src1"] = "/World/Actor1"
        tritri_dict["stiffness"] = float(1.0)
        tritri_dict["damping"] = float(10.0)
        self.expected_prims["/World/TriTriAttachment" + "/triTriAttachment"] = tritri_dict

        await self.parse("Attachment")


    async def test_collision_filter(self):
        self.fail_on_log_error = True
        collision_filter_dict = {}
        self.expected_prims = {}

        collision_filter_dict["enabled"] = True
        collision_filter_dict["src0"] = "/World/Actor0"
        collision_filter_dict["src1"] = "/World/Actor1"
        self.expected_prims["/World/CollisionFilter" + "/collisionFilter"] = collision_filter_dict

        await self.parse("Attachment")

File: omni/extensions/runtime/source/omni.usdphysics/plugins/Attachment.cpp
// SPDX-FileCopyrightText: Copyright (c) 2019-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause
//

// This include must come first
// clang-format off
#include "UsdPCH.h"
// clang-format on

#include <carb/Types.h>
#include <carb/logging/Log.h>
#include <private/omni/physics/schema/IUsdPhysicsListener.h>

#include <common/utilities/Utilities.h>
#include <common/utilities/UsdMaterialParsing.h>

#include "Attachment.h"
#include "UsdLoad.h"

using namespace pxr;

namespace omni
{
    namespace physics
    {
        namespace schema
        {

            bool getSourcePath(const UsdRelationship& srcRel, SdfPath& srcPath)
            {
                if (srcRel)
                {
                    SdfPathVector sources;
                    srcRel.GetTargets(&sources);
                    if (sources.size() == 1)
                    {
                        srcPath = sources[0];
                        return true;
                    }
                    else
                    {
                        CARB_LOG_ERROR("Invalid number of target sources (%zd). Only 1 target source is allowed for attachments.", sources.size());
                    }
                }

                return false;
            }

            AttachmentDesc* parseAttachment(const UsdPrim prim, const ObjectType::Enum objectType, const TfToken derivedTypeName)
            {
                const TfType derivedType = UsdSchemaRegistry::GetTypeFromSchemaTypeName(derivedTypeName);
                if (!prim.IsA(derivedType))
                {
                    return nullptr;
                }

                AttachmentDesc* desc = new AttachmentDesc(objectType);
                if (!desc)
                {
                    return nullptr;
                }

                prim.GetAttribute(OmniPhysicsDeformableAttrTokens->attachmentEnabled).Get(&desc->enabled);

                UsdRelationship src0Rel = prim.GetRelationship(OmniPhysicsDeformableAttrTokens->src0);
                UsdRelationship src1Rel = prim.GetRelationship(OmniPhysicsDeformableAttrTokens->src1);
                bool result = true;
                result &= getSourcePath(src0Rel, desc->src0);
                result &= getSourcePath(src1Rel, desc->src1);
                if (!result)
                {
                    delete desc;
                    return nullptr;
                }

                prim.GetAttribute(OmniPhysicsDeformableAttrTokens->damping).Get(&desc->damping);
                prim.GetAttribute(OmniPhysicsDeformableAttrTokens->stiffness).Get(&desc->stiffness);

                return desc;
            }

            AttachmentDesc* parseAttachment(const UsdStageWeakPtr stage, const UsdPrim& prim, uint64_t typeFlags)
            {
                TfType aType = UsdSchemaRegistry::GetTypeFromSchemaTypeName(OmniPhysicsDeformableTypeTokens->Attachment);
                if (!prim.IsA(aType))
                {
                    return nullptr;
                }

                if (typeFlags & PrimType::eUsdPhysicsVtxVtxAttachment)
                {
                    return parseAttachment(prim, ObjectType::eAttachmentVtxVtx, OmniPhysicsDeformableTypeTokens->VtxVtxAttachment);
                }
                else if (typeFlags & PrimType::eUsdPhysicsVtxTriAttachment)
                {
                    return parseAttachment(prim, ObjectType::eAttachmentVtxTri, OmniPhysicsDeformableTypeTokens->VtxTriAttachment);
                }
                else if (typeFlags & PrimType::eUsdPhysicsVtxTetAttachment)
                {
                    return parseAttachment(prim, ObjectType::eAttachmentVtxTet, OmniPhysicsDeformableTypeTokens->VtxTetAttachment);
                }
                else if (typeFlags & PrimType::eUsdPhysicsVtxCrvAttachment)
                {
                    return parseAttachment(prim, ObjectType::eAttachmentVtxCrv, OmniPhysicsDeformableTypeTokens->VtxCrvAttachment);
                }
                else if (typeFlags & PrimType::eUsdPhysicsVtxXformAttachment)
                {
                    return parseAttachment(prim, ObjectType::eAttachmentVtxXform, OmniPhysicsDeformableTypeTokens->VtxXformAttachment);
                }
                else if (typeFlags & PrimType::eUsdPhysicsTetXformAttachment)
                {
                    return parseAttachment(prim, ObjectType::eAttachmentTetXform, OmniPhysicsDeformableTypeTokens->TetXformAttachment);
                }
                else if (typeFlags & PrimType::eUsdPhysicsTriTriAttachment)
                {
                    return parseAttachment(prim, ObjectType::eAttachmentTriTri, OmniPhysicsDeformableTypeTokens->TriTriAttachment);
                }
                return nullptr;
            }

            ElementCollisionFilterDesc* parseCollisionFilter(const UsdStageWeakPtr stage, const UsdPrim& prim)
            {
                TfType fType = UsdSchemaRegistry::GetTypeFromSchemaTypeName(OmniPhysicsDeformableTypeTokens->ElementCollisionFilter);
                if (!prim.IsA(fType))
                {
                    return nullptr;
                }

                ElementCollisionFilterDesc* desc = new ElementCollisionFilterDesc();

                prim.GetAttribute(OmniPhysicsDeformableAttrTokens->filterEnabled).Get(&desc->enabled);

                UsdRelationship src0Rel = prim.GetRelationship(OmniPhysicsDeformableAttrTokens->src0);
                UsdRelationship src1Rel = prim.GetRelationship(OmniPhysicsDeformableAttrTokens->src1);
                bool result = true;
                result &= getSourcePath(src0Rel, desc->src0);
                result &= getSourcePath(src1Rel, desc->src1);
                if (!result)
                {
                    delete desc;
                    return nullptr;
                }

                return desc;
            }

        } // namespace schema
    } // namespace physics
} // namespace omni
